haystack/preview/components/generators/openai.py

```diff
@@ -0,0 +1,282 @@
+
+    def _get_telemetry_data(self) -> Dict[str, Any]:
+        """
+        Data that is sent to Posthog for usage analytics.
+        """
+        return {"model": self.model_name}
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        :return: The serialized component as a dictionary.
+        """
+        callback_name = serialize_callback_handler(self.streaming_callback) if self.streaming_callback else None
+        return default_to_dict(
+            self,
+            model_name=self.model_name,
+            streaming_callback=callback_name,
+            api_base_url=self.api_base_url,
+            **self.generation_kwargs,
+            system_prompt=self.system_prompt,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "GPTGenerator":
+        """
+        Deserialize this component from a dictionary.
+        :param data: The dictionary representation of this component.
+        :return: The deserialized component instance.
+        """
+        init_params = data.get("init_parameters", {})
+        serialized_callback_handler = init_params.get("streaming_callback")
+        if serialized_callback_handler:
+            data["init_parameters"]["streaming_callback"] = deserialize_callback_handler(serialized_callback_handler)
+        return default_from_dict(cls, data)
+
+    @component.output_types(replies=List[str], metadata=List[Dict[str, Any]])
+    def run(self, prompt: str, generation_kwargs: Optional[Dict[str, Any]] = None):
+        """
+        Invoke the text generation inference based on the provided messages and generation parameters.
+
+        :param prompt: The string prompt to use for text generation.
+        :param generation_kwargs: Additional keyword arguments for text generation. These parameters will
+        potentially override the parameters passed in the __init__ method.
+        For more details on the parameters supported by the OpenAI API, refer to the
+        OpenAI [documentation](https://platform.openai.com/docs/api-reference/chat/create).
+        :return: A list of strings containing the generated responses and a list of dictionaries containing the metadata
+        for each response.
+        """
+        message = ChatMessage.from_user(prompt)
+        if self.system_prompt:
+            messages = [ChatMessage.from_system(self.system_prompt), message]
+        else:
+            messages = [message]
+
+        # update generation kwargs by merging with the generation kwargs passed to the run method
+        generation_kwargs = {**self.generation_kwargs, **(generation_kwargs or {})}
+
+        openai_chat_message_format = ["role", "content", "name"]
+        openai_formatted_messages = [
+            dataclasses.asdict(
+                m, dict_factory=lambda obj: {k: v for k, v in obj if k in openai_chat_message_format and v}
+            )
+            for m in messages
+        ]
```

Comment:
This part seems a bit complex and perhaps too compact (the dict comprehension inside a list comprehension).

I would like it to be more explicit (maybe in a external method?)

Reply:
Ok will do